### PR TITLE
[SPARK-23781][CORE] Merge token renewer functionality into HadoopDelegationTokenManager.

### PR DESCRIPTION
**Pull request by vanzin**
This avoids having two classes to deal with tokens; now the above
class is a one-stop shop for dealing with delegation tokens. The
YARN backend extends that class instead of doing composition like
before, resulting in a bit less code there too.

The renewer functionality is basically the same code that used to
be in YARN's AMCredentialRenewer. That is also the reason why the
public API of HadoopDelegationTokenManager is a little bit odd;
the YARN AM has some odd requirements for how this all should be
initialized, and the weirdness is needed currently to support that.

Tested:
- YARN with stress app for DT renewal
- Mesos and K8S with basic kerberos tests (both tgt and keytab)

### CODE REVIEW

**Comment by vanzin** [comment]
I'm trying to get things to work on Mesos in my hacked cluster, might need some small adjustments to the code.

**Comment by ifilonenko** [comment]
If we are talking about the token renewal functionality, could we possibly refactor the `HadoopFSDelegationTokenProvider` as well. I found that within the function `obtainDelegationTokens()`:

This code-block:
```
    val fetchCreds = fetchDelegationTokens(getTokenRenewer(hadoopConf), fsToGetTokens, creds)

    // Get the token renewal interval if it is not set. It will only be called once.
    if (tokenRenewalInterval == null) {
      tokenRenewalInterval = getTokenRenewalInterval(hadoopConf, sparkConf, fsToGetTokens)
    }
```
calls `fetchDelegationTokens()` twice since the `tokenRenewalInterval` will always be null upon creation of the `TokenManager` which I think is unnecessary in the case of Kubernetes (as you are creating 2 DTs when only one is needed.) Idk if use-case is different in Mesos / Yarn, but could this possibly be refactored to only call `fetchDelegationTokens()` once upon startup or to have a param to specify `tokenRenewalInterval`? I could send a follow-up PR if desired, but idk if this fits better within the scope of this PR.

**Comment by vanzin** [comment]
> ...calls fetchDelegationTokens() twice since the tokenRenewalInterval...

That is true - the second fetch is only needed for YARN because reasons. But that's also completely unrelated to this change, so please file a separate bug for it.

**Comment by vanzin** [comment]
Given the lack of reviews let me ping some people: @skonto @jerryshao

File: core/src/main/scala/org/apache/spark/deploy/security/AbstractCredentialRenewer.scala
Comment (ifilonenko) [inline core/src/main/scala/org/apache/spark/deploy/security/AbstractCredentialRenewer.scala]: Should a `require(driver.isDefined)` be thrown here to ensure that the driver sends the updates?

**Review by ifilonenko — COMMENTED** [review]
LGTM, pending successful testing on Mesos @skonto. I like the abstraction and its centralization of logic and confs.

File: core/src/main/scala/org/apache/spark/deploy/security/AbstractCredentialRenewer.scala
Comment (vanzin) [reply core/src/main/scala/org/apache/spark/deploy/security/AbstractCredentialRenewer.scala]: YARN starts the token renewer before there is a driver reference available; the AM gets that later (when it connects back to the driver in client mode, or when the user code is started in cluster mode).

I've been trying to come up with a better API here that would avoid these checks, but haven't figure it out yet.

Basically for the YARN case you would need:

- an initial call to login and fetch delegation tokens, before the driver is even known (the fix for SPARK-23361)
- a call to start the renewal task, which needs information retrieved in the above call (the renewal interval)

So breaking it into two calls would need some state to be kept between the calls. Not sure whether that would be much cleaner than what's here now.

Trying to follow the Mesos way (where the driver itself manages the renewer, instead of the AM code in YARN) would probably not work in YARN, because the AM tries to connect to HDFS (for distributed cache functionality) before the driver is running, so it needs those tokens. It might be possible to rework some of the YARN backend internals to avoid that, I think, but it needs more thought and would be better left as a separate / future enhancement. 

At some point I'd also like to change the client mode on YARN to be more like the Mesos case (where the driver manages the login, not the AM), maybe all this could be cleaned up at that time.

File: resource-managers/mesos/src/main/scala/org/apache/spark/scheduler/cluster/mesos/MesosCoarseGrainedSchedulerBackend.scala
Comment (attilapiros) [inline resource-managers/mesos/src/main/scala/org/apache/spark/scheduler/cluster/mesos/MesosCoarseGrainedSchedulerBackend.scala]: Option?

File: core/src/main/scala/org/apache/spark/deploy/security/AbstractCredentialRenewer.scala
Comment (attilapiros) [reply core/src/main/scala/org/apache/spark/deploy/security/AbstractCredentialRenewer.scala]: What about extending this start() method with a boolean flag sendingInitialToken so it would be clear why this functionality is here and also this would not be controlled by the having driver registered beforehand (would be easy to see for Mesos this is a must have so the require can be used too in the if body).

File: core/src/main/scala/org/apache/spark/deploy/security/AbstractCredentialRenewer.scala
Comment (attilapiros) [inline core/src/main/scala/org/apache/spark/deploy/security/AbstractCredentialRenewer.scala]: why reading this config value every time? (As I have seen other places this is extracted to member vals).

File: core/src/main/scala/org/apache/spark/deploy/security/AbstractCredentialRenewer.scala
Comment (vanzin) [reply core/src/main/scala/org/apache/spark/deploy/security/AbstractCredentialRenewer.scala]: > What about extending this start() method with a boolean flag sendingInitialToken

That is just as awkward, because then you'd have to set the driver ref before calling start with that flag set to true.

The API is just going to be a little awkward until we refactor the way the YARN code deals with tokens. The goal here is not to fix the API, but to share all the rest of the logic.

File: resource-managers/mesos/src/main/scala/org/apache/spark/scheduler/cluster/mesos/MesosHadoopDelegationTokenManager.scala
Comment (vanzin) [reply resource-managers/mesos/src/main/scala/org/apache/spark/scheduler/cluster/mesos/MesosHadoopDelegationTokenManager.scala]: That's the existing name and I'd rather not change it.

File: core/src/main/scala/org/apache/spark/deploy/security/AbstractCredentialRenewer.scala
Comment (vanzin) [reply core/src/main/scala/org/apache/spark/deploy/security/AbstractCredentialRenewer.scala]: `Option` doesn't add anything useful here.

I want `createAndUpdateTokens` here because it will be useful for k8s client mode and for when the YARN code is refactored. Just like it is in Mesos today.

File: core/src/main/scala/org/apache/spark/deploy/security/AbstractCredentialRenewer.scala
Comment (vanzin) [reply core/src/main/scala/org/apache/spark/deploy/security/AbstractCredentialRenewer.scala]: This makes it more flexible. I also really dislike returning tuples.

File: core/src/main/scala/org/apache/spark/deploy/security/AbstractCredentialRenewer.scala
Comment (vanzin) [reply core/src/main/scala/org/apache/spark/deploy/security/AbstractCredentialRenewer.scala]: The goal is to make it simple for scheduler backends to use token functionality. Adding more classes that the backends need to deal with complicates things instead of making them simpler.

File: resource-managers/mesos/src/main/scala/org/apache/spark/scheduler/cluster/mesos/MesosCoarseGrainedSchedulerBackend.scala
Comment (vanzin) [reply resource-managers/mesos/src/main/scala/org/apache/spark/scheduler/cluster/mesos/MesosCoarseGrainedSchedulerBackend.scala]: Just adds unnecessary indirection.

File: core/src/main/scala/org/apache/spark/deploy/security/AbstractCredentialRenewer.scala
Comment (vanzin) [reply core/src/main/scala/org/apache/spark/deploy/security/AbstractCredentialRenewer.scala]: What is the advantage to stashing it in a variable? It just decouples the code that is reading the config from the code that is using it.

**Comment by vanzin** [comment]
I merged start() with createAndUpdateTokens() in the last commit; the API is still awkward because of YARN's usage, but it simplifies the Mesos code a little bit. It can be easily reverted in any case.

Tested on Mesos, testing on YARN at the moment.

File: resource-managers/mesos/src/main/scala/org/apache/spark/scheduler/cluster/mesos/MesosHadoopDelegationTokenManager.scala
Comment (attilapiros) [inline resource-managers/mesos/src/main/scala/org/apache/spark/scheduler/cluster/mesos/MesosHadoopDelegationTokenManager.scala]: What about `driver.isDefined` instead of `driver.orNull != null`?

Comment (vanzin) [reply resource-managers/mesos/src/main/scala/org/apache/spark/scheduler/cluster/mesos/MesosHadoopDelegationTokenManager.scala]: `Some(null).isDefined` is true. And if you get initialization order wrong in these classes, you can get into that situation. So this is safer.

**Comment by vanzin** [comment]
Another small change to rename AbstractCredentialRenewer -> AbstractCredentialManager. I prefer that name better since I'm adding more than just renewal functionality to that class, but it sort of clashes with the existing classes. I'll file a separate bug to rename  those classes instead.

**Comment by vanzin** [comment]
Trying more pings: @jerryshao @tgravescs

File: core/src/main/scala/org/apache/spark/deploy/security/HadoopDelegationTokenManager.scala
Comment (ifilonenko) [inline core/src/main/scala/org/apache/spark/deploy/security/HadoopDelegationTokenManager.scala]: Are there going to be unit tests for this?  Because trying to mock lines like `UserGroupInformation.getCurrentUser().getCredentials()` will be tricky. If not, then the way the methods are structured appear to be fine, given successful e2e tests.

File: core/src/main/scala/org/apache/spark/deploy/security/HadoopDelegationTokenManager.scala
Comment (ifilonenko) [inline core/src/main/scala/org/apache/spark/deploy/security/HadoopDelegationTokenManager.scala]: Can these tokens be packaged in a tuple or case class with the UGI? i.e. 
`(tokens, ugi)` as a response for this method 
Reason being, (I know it is preferred to use driver.send(token)), but logic in Kubernetes makes it simple to merely update the secret (as updates to Secrets are automatically detected by Driver + Executor pods and this logic is already being run natively by the k8s api). Unless we are being strict in being idiomatic to Spark, I think leveraging the benefits of Kubernetes here would be beneficial. I'll leave it up to you in deciding how executors should detect a newly updated token.

**Review by ifilonenko — COMMENTED** [review]
Just sharing opinions on the side of Kubernetes, w.r.t how executors get the new tokens. 

Also, if it is desired to unit-test, some of the methods need to be refactored for mocking purposes, I believe.

File: core/src/main/scala/org/apache/spark/deploy/security/HadoopDelegationTokenManager.scala
Comment (vanzin) [reply core/src/main/scala/org/apache/spark/deploy/security/HadoopDelegationTokenManager.scala]: This code has always lacked proper unit tests. In part it's because it's hard to tell what a good unit test is here, in part because no one ever bothered. It's also lacking integration tests, although I have a few internally that I generally run on these changes.

That needs to be looked at separately; adding changes here without knowing what a good unit test for this code is would be kinda pointless.

File: core/src/main/scala/org/apache/spark/deploy/security/HadoopDelegationTokenManager.scala
Comment (vanzin) [reply core/src/main/scala/org/apache/spark/deploy/security/HadoopDelegationTokenManager.scala]: > makes it simple to merely update the secret 

The problem is that you still need Spark to load the updated secrets, and where is that logic?

I'd rather keep all backends working the same way unless there is a good reason for them not to. If there is a good reason, then later you can make this change when you actually write code to implement that. But right now I don't see a reason for it.

Comment (ifilonenko) [reply core/src/main/scala/org/apache/spark/deploy/security/HadoopDelegationTokenManager.scala]: > The problem is that you still need Spark to load the updated secrets, and where is that logic?

Agreed, it isn't there yet. But I agree, such a change could be added on later, but it will just require a slight refactor of this method which I wanted to pre-empt. But sounds good!

File: core/src/main/scala/org/apache/spark/deploy/security/HadoopDelegationTokenManager.scala
Comment (ifilonenko) [inline core/src/main/scala/org/apache/spark/deploy/security/HadoopDelegationTokenManager.scala]: Can we make this method overrideable since, despite these two cases being the only ones in Yarn and Mesos, Kubernetes does have a third option with pre-defined secrets, using Kubernetes specific Watchers, I wish to leverage that will also end with `ref.send(UpdateDelegationTokens(tokens)`? Unless you prefer to have the follow-up PR do this refactoring.

Comment (vanzin) [reply core/src/main/scala/org/apache/spark/deploy/security/HadoopDelegationTokenManager.scala]: Yes, I'd prefer any k8s-specific enhancements to be made together with the corresponding k8s code.

**Review by ifilonenko — APPROVED** [review]
This passes all Kubernetes integration tests (including Kerberos), so it is a +1 from me. @skonto if you could weigh in on the Mesos side, would be great. Otherwise, this looks good to merge.

File: resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/k8s/security/KubernetesHadoopDelegationTokenManager.scala
Comment (squito) [inline resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/k8s/security/KubernetesHadoopDelegationTokenManager.scala]: this class doesn't really seem necessary anymore, but not a big deal

File: core/src/main/scala/org/apache/spark/deploy/security/HadoopDelegationTokenManager.scala
Comment (squito) [inline core/src/main/scala/org/apache/spark/deploy/security/HadoopDelegationTokenManager.scala]: `driver` is also the name of the arg to this method, can you rename one or the other?  maybe rename the arg to `driverOpt`?

File: core/src/main/scala/org/apache/spark/deploy/security/HadoopDelegationTokenManager.scala
Comment (squito) [inline core/src/main/scala/org/apache/spark/deploy/security/HadoopDelegationTokenManager.scala]: and would nice to use the same naming as above here as well, eg. maybe `driverOpt.foreach { driver =>`

**Review by squito — COMMENTED** [review]
I only really looked at the yarn code, and it took me a couple reads to understand what was going on.  IIUC, this sets things up for token renewal to happen inside the `CoarseGrainedSchedulerBackend`.  I guess mesos makes use of that.  However yarn does not use that -- its still happening in the AM, and the AM sets the driverRef on the tokenManager.  That driverRef is actually YarnScheduler.YarnSchedulerEndpoint, which will relay the message to CoarseGrainedSchedulerBackend.DriverEndpoint.

At first I thought this was making a bigger change to move the renewal from the AM into the driver.

Is there any place which documents this overall flow?  I couldn't find one, and I think it would be good to have a place.  Maybe YarnHadoopDelegationTokenManager could describe how its not meant to live inside the scheduler, but instead inside the AM.

Also I'm wondering, do you want to add delegations tokens to happen inside CoarseGrainedSchedulerBackend instead of YarnSchedulerBackend?  doesn't seem yarn specific:

https://github.com/apache/spark/blob/24e8c27dfe31e6e0a53c89e6ddc36327e537931b/resource-managers/yarn/src/main/scala/org/apache/spark/scheduler/cluster/YarnSchedulerBackend.scala#L273

File: core/src/main/scala/org/apache/spark/deploy/security/HadoopDelegationTokenManager.scala
Comment (vanzin) [reply core/src/main/scala/org/apache/spark/deploy/security/HadoopDelegationTokenManager.scala]: I actually dislike that this check exists in so many places... but sure I'll make it consistent.

File: core/src/main/scala/org/apache/spark/deploy/security/HadoopDelegationTokenManager.scala
Comment (vanzin) [reply core/src/main/scala/org/apache/spark/deploy/security/HadoopDelegationTokenManager.scala]: I think `NoClassDefFoundError` is the main one but I'm not sure whether others may occur. I think SO or OOM here are very unlikely so it's a pretty small risk to leave this as is.

File: resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/k8s/security/KubernetesHadoopDelegationTokenManager.scala
Comment (vanzin) [reply resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/k8s/security/KubernetesHadoopDelegationTokenManager.scala]: This class (and a whole lot of code on the k8s side) is actually in my crosshairs for further cleanup; but that would mean changes that are unrelated to this one, so separate.

**Comment by vanzin** [comment]
> Is there any place which documents this overall flow?

Not yet. I plan to write a README.md explaining this, separately, since there's a lot of
confusion not just about the flow, but about exactly what this code does and why.

> add delegations tokens inside CoarseGrainedSchedulerBackend instead of YarnSchedulerBackend

I'm not 100% sure that's necessary. In YARN client mode + keytab, it's needed because
the AM handles the login and sends DTs back to the driver. But in other modes, the driver
is expected to have a valid TGT at all times, so it should not need DTs.

But I see that the old Mesos code did what you're saying, so perhaps it's better to keep
that and avoid a regression. I just need to test that doesn't do anything funny on YARN.

File: core/src/main/scala/org/apache/spark/deploy/security/HadoopDelegationTokenManager.scala
Comment (squito) [reply core/src/main/scala/org/apache/spark/deploy/security/HadoopDelegationTokenManager.scala]: yeah agree its weird that the check is in so many places -- that bugged me too but I wasn't really sure about whether there were other entry points.  but in any case, yes I think anyplace we're going to check, lets be consistent.

**Comment by squito** [comment]
>> Is there any place which documents this overall flow?
>
> Not yet. I plan to write a README.md explaining this, separately, since there's a lot of
confusion not just about the flow, but about exactly what this code does and why.

OK sounds great.  yeah would be really good to have something explaining the different scenarios, I'm often worried when I'm looking at this code I'm forgetting about some weird case (eg. the AM restarts after 7 days but everything else stays up, etc.)

**Comment by vanzin** [comment]
I tested the last update with yarn cluster and client, both seem happy.

File: core/src/main/scala/org/apache/spark/scheduler/cluster/CoarseGrainedSchedulerBackend.scala
Comment (squito) [inline core/src/main/scala/org/apache/spark/scheduler/cluster/CoarseGrainedSchedulerBackend.scala]: realized the other changes were small so I took a look -- this is the only part which concerns me.  This used to be a blocking call, so you were sure that you'd get tokens to respond to `RetrieveSparkAppConfig`.  However, now you're getting the delegation tokens async.  You will probably get them back from the TokenManager before sending them here, but there is a race, right?  Will it be OK if the executor doesn't get the tokens here?  I know you'll send the executor the tokens in an `UpdateDelegationTokens` msg later, but will that lead to some spurious failures?

you could change the token manager to call `driver.ask(UpdateDelegationTokens(tokens))` inside `start()`, and ensure that future is completed here, though it makes things more complex.

File: core/src/main/scala/org/apache/spark/deploy/security/HadoopDelegationTokenManager.scala
Comment (vanzin) [reply core/src/main/scala/org/apache/spark/deploy/security/HadoopDelegationTokenManager.scala]: This is the part of the current API that I dislike the most, but I really don't know how to make it better without fixing SPARK-25689 first.

This at least keeps a single method to initialize delegation tokens, instead of having logic in the call sites about what method to call.

File: core/src/main/scala/org/apache/spark/scheduler/cluster/CoarseGrainedSchedulerBackend.scala
Comment (vanzin) [reply core/src/main/scala/org/apache/spark/scheduler/cluster/CoarseGrainedSchedulerBackend.scala]: Although it's not obvious I think this is fine.

The token manager is started before any rm-specific code runs and requests executors (in the `CoarseGrainedSchedulerBackend.start()` method; rm-specific schedulers call that before they do anything). So the message to set these tokens is the first message queued on the driver, and this reference will already be set when the first executor asks for this info.

Comment (vanzin) [reply core/src/main/scala/org/apache/spark/scheduler/cluster/CoarseGrainedSchedulerBackend.scala]: Actually since I moved the token manager to `CoarseGrainedSchedulerBackend` I think I can simplify this and avoid having to trace how the RPC code works to prove this is fine. I'll try to update things this afternoon after I finish some other stuff.

Comment (squito) [reply core/src/main/scala/org/apache/spark/scheduler/cluster/CoarseGrainedSchedulerBackend.scala]: Not sure I follow.  to be clear, I definitely think this is extremely unlikely, but I don't see what totally prevents it.

>  So the message to set these tokens is the first message queued on the driver, and this reference will already be set when the first executor asks for this info.

I see that the message will get queued from the TokenManager to get sent to the SchedulerBackend before anything else happens.  But I was a bit hazy on what happens after that -- I guess because both are local, the way the rpc env works, nothing else can squeeze in-between?  its definitely not obvious -- seems its worth at least a short comment like "Because the  token manager is local, we know we will have already received the initial set of tokens (if any)"

File: core/src/main/scala/org/apache/spark/deploy/security/HadoopDelegationTokenManager.scala
Comment (squito) [reply core/src/main/scala/org/apache/spark/deploy/security/HadoopDelegationTokenManager.scala]: ok so we agree that is a bit ugly and maybe there isn't a better api currently -- but are you opposed to the other suggestions, till the other changes happen?  I know its a small thing, I just think it would help keep the usage clear (or another set of asserts to resolve the ambiguity another way is fine too, I just don't like the ambiguity).

File: core/src/main/scala/org/apache/spark/scheduler/cluster/CoarseGrainedSchedulerBackend.scala
Comment (vanzin) [reply core/src/main/scala/org/apache/spark/scheduler/cluster/CoarseGrainedSchedulerBackend.scala]: >  But I was a bit hazy on what happens after that

The RPC layer is basically a message queue. So first message to be sent = first message to be processed.

But as I said, I'll clean this up since now the code is all in `CoarseGrainedSchedulerBackend` and that makes it easier.

**Review by squito — APPROVED** [review]
lgtm
thanks for the last changes, I think that makes it easier to follow

**Comment by squito** [comment]
any more comments @ifilonenko @skonto ?

**Comment by ifilonenko** [comment]
For Kerberos on K8s this looks good, LGTM from me.
I have already begun building a PR off of this work. and it seems to be quite easy to work with.

**Comment by squito** [comment]
merged to master

**Comment by ifilonenko** [comment]
Thank you for this work @vanzin !

**Pull request closed without merging**
